init rolling file appender

urso opened the pull request.

Commit:
init rolling file appender

Commit:
progress

michalpristas commented on backend/appender/rolling/rolling.go:
```diff
@@ -0,0 +1,178 @@
+}
+
+// Rolloverer is used by the trigger to start the rollover process.
+type Rolloverer interface {
```
can we come up with a better name. I think `-er` rule is not always the best option. I would settle with simple `Roller`, `LogRoller`, `FileRoller` or `FileRotator`
This is just a small nit but i think it slightly improves readability especially with exported interfaces

urso replied:
I renamed it to 'Rotator'.

Commit:
rename rolloverer to rotator

Commit:
number of small changes

- rename rolling to Appender and export symbol (full name
  rolling.Appender)
- add support to close the rolling appender
- add Background type to track and shutdown background processes
  triggers and strategies might want to run.

Commit:
finish triggers

Commit:
rename file

Commit:
continue file rotation

Commit:
minor fixes

urso commented:
@michalpristas 	@ph The rolling file appender is basically working (the whole project needs some tests :) ). Only thing I wonder is: Do we want to be able to configure a limit on log message sizes (bigger ones will be filtered out)? If so, how to configure it? Right now the file size is configured via 'trigger' (which checks the size after writing the log message).

I include the data in the rotated names, this reduces the number of rename operations we have to run, removes the file deletion problems with external collectors (like filebeat) on windows.

Commit:
fix size counter on newline

ph commented:
> @michalpristas @ph The rolling file appender is basically working (the whole project needs some tests :) ). Only thing I wonder is: Do we want to be able to configure a limit on log message sizes (bigger ones will be filtered out)? If so, how to configure it? Right now the file size is configured via 'trigger' (which checks the size after writing the log message).

Lets get this merge and we can create another issue.

Sorry I've missed the notification somehow.

michalpristas commented on backend/appender/rolling/appender.go:
```diff
@@ -0,0 +1,210 @@
+}
+
+func (a *Appender) Write(b []byte) (int, error) {
+	if a.file == nil {
```
exported method, is it safe to omit lock here? this possibly calls rotate

urso replied:
The writer is not called directly from the process, but it is called from the 'layout'. That is the lock is already taken. Will try to make this method 'unexported'.

urso replied:
Did update the code such that Write is not exported anymore (plus added some comments).

Commit:
Do not export (*Appender).Write

ph commented:
Thanks @urso
